### ISSUE TITLE
Isolator deadlocks with plugin process

### ISSUE THREAD

**Issue by spikecurtis** [mesosphere/net-modules#81]
I've seen an issue while scale testing where the `Isolator` deadlocks with the plug-in binary.

The `calico_mesos` plug in comes up, and logs out that it is starting, then does a `sys.stdin.read()` call.  This appears to be where we are deadlocking.  Python's `read()` waits for `EOF` before returning, so my suspicion is that sometimes Mesos fails to send it.

It happens only intermittently, so must be a resource exhaustion or race condition problem.

This has serious knock-on issues as the entire agent process gets deadlocked, master pings time out and the entire agent is declared lost.

**Comment by spikecurtis** [comment]
cc @robbrockbank @djosborne

**Comment by robbrockbank** [comment]
I made some modification to the calico_mesos plugin to read a byte at a time and to timeout after 2s if there is no data.  Looks like we are getting the first byte `{` succesffuly, but are then timing out waiting for more data.

Could it be that the stdin pipe from the sub-process needs to be flushed?  I could give that a try if someone points me at the correct incantation to do that.

**Comment by robbrockbank** [comment]
Another update.  I tested with a modified version of network_isolator.cpp and the calico_mesos binary as follows:
-  I completely removed the section of network_isolator.cpp that closed stdin, and added code to append a new line to the json message.
-  I updated the calico_mesos executable to parse and process the json message when it receives a newline - so it does not wait for stdin to close.

With these changes I did not hit the original deadlock - so it really does seem to be a window condition to do with how stdin is closed.
